pinot-core/src/main/java/org/apache/pinot/core/query/reduce/GapFillGroupByDataTableReducer.java

```diff
@@ -0,0 +1,483 @@
+          new PostAggregationHandler(_queryContext, prePostAggregationDataSchema);
+      FilterContext havingFilter = _queryContext.getHavingFilter();
+      if (havingFilter != null) {
+        HavingFilterHandler havingFilterHandler = new HavingFilterHandler(havingFilter, postAggregationHandler);
+        while (rows.size() < limit && sortedIterator.hasNext()) {
+          Object[] row = sortedIterator.next().getValues();
+          extractFinalAggregationResults(row);
+          for (int i = 0; i < numColumns; i++) {
+            row[i] = columnDataTypes[i].convert(row[i]);
+          }
+          if (havingFilterHandler.isMatch(row)) {
+            rows.add(row);
+          }
+        }
+      } else {
+        for (int i = 0; i < limit && sortedIterator.hasNext(); i++) {
+          Object[] row = sortedIterator.next().getValues();
+          extractFinalAggregationResults(row);
+          for (int j = 0; j < numColumns; j++) {
+            row[j] = columnDataTypes[j].convert(row[j]);
+          }
+          rows.add(row);
+        }
+      }
+      DataSchema resultDataSchema = postAggregationHandler.getResultDataSchema();
+      ColumnDataType[] resultColumnDataTypes = resultDataSchema.getColumnDataTypes();
+      List<Object[]> resultRows = new ArrayList<>(rows.size());
+      for (Object[] row : rows) {
+        Object[] resultRow = postAggregationHandler.getResult(row);
+        for (int i = 0; i < resultColumnDataTypes.length; i++) {
+          resultRow[i] = resultColumnDataTypes[i].format(resultRow[i]);
+        }
+        resultRows.add(resultRow);
+        _primaryKeys.add(constructKey(resultRow));
+      }
+      List<Object[]> gapfillResultRows = gapFill(resultRows, resultColumnDataTypes);
+      brokerResponseNative.setResultTable(new ResultTable(resultDataSchema, gapfillResultRows));
+    } else {
+      // PQL query with SQL group-by mode and response format
+      // NOTE: For PQL query, keep the order of columns as is (group-by expressions followed by aggregations), no need
+      //       to perform post-aggregation or filtering.
+
+      for (int i = 0; i < limit && sortedIterator.hasNext(); i++) {
+        Object[] row = sortedIterator.next().getValues();
+        extractFinalAggregationResults(row);
+        for (int j = 0; j < numColumns; j++) {
+          row[j] = columnDataTypes[j].convertAndFormat(row[j]);
+        }
+        rows.add(row);
+      }
+      brokerResponseNative.setResultTable(new ResultTable(prePostAggregationDataSchema, rows));
+    }
+  }
+
+  List<Object[]> gapFill(List<Object[]> resultRows, ColumnDataType[] resultColumnDataTypes) {
+    int limit = _queryContext.getLimit();
+    int numResultColumns = resultColumnDataTypes.length;
+    List<Object[]> gapfillResultRows = new ArrayList<>(limit);
+    long step = _dateTimeGranularity.granularityToMillis();
+    int index = 0;
+    for (long time = _startMs; time + 2 * step <= _endMs; time += step) {
```

Comment:
We are looping over the rows multiple times (once per time bucket). We can optimize it to only loop over the rows once. One option is to track the index of the previous iteration and start from the end index of the previous run

Reply:
We are not looping over the rows multiple times. The "index" is for resultRows and also moving forward.